library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java

```diff
@@ -0,0 +1,486 @@
+  {
+    this.dataStorage = storageAgent;
+  }
+
+  /**
+   * This method sets the storage for the retraction data for each window. Only used when the accumulation mode is ACCUMULATING_AND_RETRACTING
+   *
+   * @param storageAgent
+   */
+  public void setRetractionStorage(DataStorageT storageAgent)
+  {
+    this.retractionStorage = storageAgent;
+  }
+
+  /**
+   * Sets the accumulation, which basically tells the WindowedOperator what to do if a new tuple comes in and what
+   * to put in the pane when a trigger is fired
+   *
+   * @param accumulation
+   */
+  public void setAccumulation(AccumulationT accumulation)
+  {
+    this.accumulation = accumulation;
+  }
+
+  @Override
+  public void setWindowStateStorage(WindowedStorage<WindowState> storageAgent)
+  {
+    this.windowStateMap = storageAgent;
+  }
+
+  @Override
+  public void setTimestampExtractor(Function<InputT, Long> timestampExtractor)
+  {
+    this.timestampExtractor = timestampExtractor;
+  }
+
+  public void validate() throws ValidationException
+  {
+    if (accumulation == null) {
+      throw new ValidationException("Accumulation must be set");
+    }
+    if (dataStorage == null) {
+      throw new ValidationException("Data storage must be set");
+    }
+    if (windowStateMap == null) {
+      throw new ValidationException("Window state storage must be set");
+    }
+    if (triggerOption != null) {
+      if (triggerOption.isFiringOnlyUpdatedPanes()) {
+        if (retractionStorage == null) {
+          throw new ValidationException("A retraction storage is required for firingOnlyUpdatedPanes option");
+        }
+        if (triggerOption.getAccumulationMode() == TriggerOption.AccumulationMode.DISCARDING) {
+          throw new ValidationException("DISCARDING accumulation mode is not valid for firingOnlyUpdatedPanes option");
+        }
+      }
+      if (triggerOption.getAccumulationMode() == TriggerOption.AccumulationMode.ACCUMULATING_AND_RETRACTING &&
+          retractionStorage == null) {
+        throw new ValidationException("A retraction storage is required for ACCUMULATING_AND_RETRACTING accumulation mode");
+      }
+    }
+  }
+
+  @Override
+  public Tuple.WindowedTuple<InputT> getWindowedValue(Tuple<InputT> input)
+  {
+    Tuple.WindowedTuple<InputT> windowedTuple = new Tuple.WindowedTuple<>();
+    windowedTuple.setValue(input.getValue());
+    windowedTuple.setTimestamp(extractTimestamp(input));
+    assignWindows(windowedTuple.getWindows(), input);
+    return windowedTuple;
+  }
+
+  private long extractTimestamp(Tuple<InputT> tuple)
+  {
+    if (timestampExtractor == null) {
+      if (tuple instanceof Tuple.TimestampedTuple) {
+        return ((Tuple.TimestampedTuple)tuple).getTimestamp();
+      } else {
+        return 0;
+      }
+    } else {
+      return timestampExtractor.apply(tuple.getValue());
+    }
+  }
+
+  private void assignWindows(List<Window> windows, Tuple<InputT> inputTuple)
+  {
```

Comment:
Minor nit: Considering WindowedTuple has a addWindow method, maybe it just make sense to have only single parameter to this method "WindowedTuple<InputT> tuple".. Rest all can be derived from this object.